clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,390 @@
+            // Legacy trace support where we need the binaries.
+            std::string err = find_mapped_trace_address(trace_pc, decode_pc);
+            if (err != "")
+                return err;
+        }
+        app_pc next_pc = decode_from_copy(dcontext_, decode_pc, trace_pc, instr);
+        if (next_pc == nullptr || !instr_valid(instr)) {
+            if (persist_decoded_instrs_) {
+                instr_destroy(dcontext_, instr);
+            }
+            return "decode_from_copy failed";
+        }
+        decode_cache_[trace_pc].set_decode_info(dcontext_, memref_instr, instr);
+        return "";
+    }
+
+    /**
+     * Performs initialization tasks such as verifying whether the given trace
+     * indeed has embedded encodings or not, and initializing the
+     * \p module_mapper_t if the module path is provided.
+     *
+     * It is important to note that the trace filetype may be obtained using the
+     * get_filetype() API on a \p memtrace_stream_t. However, not all instances of
+     * \p memtrace_stream_t have the filetype available at init time (before the
+     * \p TRACE_MARKER_TYPE_FILETYPE is actually returned by the stream).
+     * Particularly, the \p stream_t provided by the \p scheduler_t to analysis
+     * tools provides the file type at init time when #dynamorio::drmemtrace::
+     * scheduler_tmpl_t::scheduler_options_t.read_inputs_in_init is set. In other
+     * cases, the user may call this API after init time when the
+     * \p TRACE_MARKER_TYPE_FILETYPE marker is seen (still before any instr
+     * record).
+     *
+     * If the provided \p module_file_path is empty, the cache object uses
+     * the encodings embedded in the trace records.
+     *
+     * If the \p module_file_path parameter is not empty, it instructs the
+     * \p decode_cache_t object that it should look for the instr encodings in the
+     * application binaries using a \p module_mapper_t. A \p module_mapper_t is
+     * instantiated only one time and reused for all objects
+     * of \p decode_cache_t (of any template type). The user must provide a valid
+     * \p module_file_path if decoding instructions from a trace that does not
+     * have embedded instruction encodings in it, as indicated by absence of
+     * the \p OFFLINE_FILE_TYPE_ENCODINGS bit in the \p TRACE_MARKER_TYPE_FILETYPE
+     * marker. The user may provide a \p module_file_path also if they
+     * deliberately need to use the module mapper instead of the embedded
+     * encodings.
+     */
+    std::string
+    init(offline_file_type_t filetype, const std::string &module_file_path = "",
+         const std::string &alt_module_dir = "")
+    {
+        // TODO i#7113: Should we perhaps also do dr_set_isa_mode for DR_ISA_REGDEPS
+        // that cannot be figured out automatically using the build but needs the
+        // trace filetype? Or does that belong in the analyzer framework which is more
+        // central than this decode_cache_t.
```

Comment:
But every tool that does decoding today currently sets the ISA to regdeps, so as-is this library can't be used in any of them; nor in any analysis tool that we want to work on the public v2 traces.

Reply:
Coming back to this: I really think this should set the isa mode.  Otherwise new analyzers on our public traces are going to all fail.

Reply:
I had made it a TODO with the intention of following up on it in a new PR (this one is already 1k+ LOC). Also, doesn't it make sense for the framework to set it? Other ISA modes (the arch-dependent ones) are set automatically by the framework today. (That change can go in first before this PR if you consider this PR incomplete without it)

Reply:
Added the ISA mode setting to init.